Realm/RLMRealm.h

```diff
@@ -500,6 +505,144 @@ typedef void (^RLMNotificationBlock)(RLMNotification notification, RLMRealm *rea
+          during a single transaction.
+ */
+@property (nonatomic, readonly) BOOL inAsyncWriteTransaction;
+
+/**
+ Begins asynchronous write transaction.
+ 
+ @param block The block containing actions to perform.
+        `block` should end by calling `commitAsyncWriteTransaction`, `cancelAsyncTransaction:`,
+        `commitWriteTransaction` or `cancelWriteTransaction`.
+        Returning without one of these calls will be equivalent to calling `cancelAsyncTransaction`.
+
+ @return The Id for the asynchronous transaction.
+ 
+ @note `block` is queued for execution on the scheduler associated with the current Realm.
+       It will run after the write transaction has begun.
+       The call returns immediately allowing the caller to proceed while the write mutex is held by someone else.
+       Write blocks from multiple calls to `beginAsyncWriteTransaction:` or `asyncTransactionWithBlock:`
+       will be executed in order.
+       A later call to `beginAsyncWriteTransaction:` or `asyncTransactionWithBlock:` will wait for any earlier
+       write blocks.
+ 
+ @note The Id for the asynchronous transaction can be used to cancel the transaction. You can get it
+       as a return from `beginAsyncWriteTransaction`, `commitAsyncWriteTransaction` or `asyncTransactionWithBlock`.
+ */
+- (RLMAsyncTransactionId)beginAsyncWriteTransaction:(void(^)())block;
+
+/** Commit asynchronous transaction.
+ 
+ The call returns immediately allowing the caller to proceed while the I/O is performed
+ on a dedicated background thread.
+ 
+ @param completionBlock  is queued for execution on the scheduler associated with
+        the current Realm. It will run after the commit has reached stable storage.
+        Callbacks to `completionBlock` will occur in the order of `commitAsyncWriteTransaction`
+ 
+ @param isGroupingAllowed If `true`, the next `commitAsyncWriteTransaction` *may* run without an
+        intervening synchronization of stable storage. Such a sequence of commits
+        form a group. It may help to have a better performance on write.
+        In case of a platform crash, either none or all of the commits in a group will reach stable storage.
+        `false` by default.
+
+ @param error If an error occurs, upon return contains an `NSError` object that describes the problem.
+        If you are not interested in possible errors, pass in `NULL`.
+
+ @return The Id for the asynchronous transaction or 0 in case of error.
+
+ @note The Id for the asynchronous transaction can be used to cancel the transaction. You can get it
+       as a return from `beginAsyncWriteTransaction`, `commitAsyncWriteTransaction` or `asyncTransactionWithBlock`.
+
+ @note The immediate return without `error` is not a guarantee of the future successful transaction commit.
+*/
+- (RLMAsyncTransactionId)commitAsyncWriteTransaction:(nullable void(^)())completionBlock isGroupingAllowed:(BOOL)isGroupingAllowed error:(NSError **)error;
```

Comment:
Why did you end up choosing an out error here instead of having the error be a block param?

Reply:
block param is better. Fixed